### PR TITLE
Feature/page title font scaling

### LINKED ISSUE #481: [Enhancement] Text size scaling dependent on amount of text
**Is your enhancement request related to a problem? Please describe.**
We would like text scaling in size on our page-title dependent on amount of text on the page-title, from default H1 to minimum in size H3.

When we build a page we would like define what our max-height is in that page (eg. 2 lines). If the page-title on this page cant be shown in that "container", we would like to scale down, before we truncate. 

Eg. if the text doesn't fit, after the max height of the page-title has been reached, the text should scale down from the original H1 to minimum H3.

**See specific examples here from inbox**
Short text: https://zpl.io/2vPrN3J
Longer text: https://zpl.io/VDo7mll
Very long text: https://zpl.io/a7vlwWE

** See generic rule with 3 lines here**

Skærmbillede 2019-10-15 kl  09 12 29

### PR DESCRIPTION
**Pull request by hafstad**
closes #481

### CODE REVIEW

File: src/kirby/directives/fit-header/fit-header.directive.ts [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+    this.observeResize();
+  }
+
+  ngOnDestroy(): void {
+    this.resizeObserverService.unobserve(this.elementRef);
+    this.renderer.removeChild(this.elementRef.nativeElement, this.clone);
+  }
+
+  private getFontSize(size): string {
+    if (
```
Comment (jakobe) [inline src/kirby/directives/fit-header/fit-header.directive.ts]: Consider simplyfing with helper method:
```
private getScssValue(path: string[]) {
    let node = this.scssVariables;
    for (let step of path) {
      node = node[step];
      if (!node) {
        break;
      }
    }
    return node;
  }
```
and use like:
`const fontSize = this.getScssValue(['global', '$font-sizes', 'value', size]);`

File: src/kirby/directives/fit-header/fit-header.directive.ts [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+  }
+
+  private getLineHeight(size): string {
+    if (
+      this.scssVariables['global'] &&
+      this.scssVariables['global']['$line-height'] &&
+      this.scssVariables['global']['$line-height']['value'] &&
+      this.scssVariables['global']['$line-height']['value'][size]
+    ) {
+      const fontSize = this.scssVariables['global']['$line-height']['value'][size];
```
Comment (jakobe) [inline src/kirby/directives/fit-header/fit-header.directive.ts]: Copy-paste typo:
```suggestion
      const lineHeight = this.scssVariables['global']['$line-height']['value'][size];
```

File: src/kirby/directives/fit-header/fit-header.directive.ts [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+@Directive({
+  selector: `h1[kirbyFitHeader],h2[kirbyFitHeader],h3[kirbyFitHeader]`,
+})
+export class FitHeaderDirective implements OnInit, OnDestroy {
+  // tslint:disable-next-line:no-input-rename
+  @Input('kirbyFitHeader') config?: FitHeaderConfig = {
+    maxLines: 2,
+  };
+
+  private lineHeight: number;
```
Comment (jakobe) [inline src/kirby/directives/fit-header/fit-header.directive.ts]: Consider inlining in `shouldScale` as it's only used there

File: src/kirby/directives/fit-header/fit-header.directive.ts [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+  private height: number;
+  private width: number;
```
Comment (jakobe) [inline src/kirby/directives/fit-header/fit-header.directive.ts]: Consider naming:
```suggestion
  private previousWidth: number;
```

File: src/kirby/directives/fit-header/fit-header.directive.ts [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+      this.scssVariables['global']['$font-sizes']['value'] &&
+      this.scssVariables['global']['$font-sizes']['value'][size]
+    ) {
+      const fontSize = this.scssVariables['global']['$font-sizes']['value'][size];
+      return `${fontSize.value}${fontSize.unit}`;
+    }
+  }
+
+  private getLineHeight(size): string {
+    if (
```
Comment (jakobe) [inline src/kirby/directives/fit-header/fit-header.directive.ts]: Consider simplyfing with helper method:
`const lineHeight = this.getScssValue(['global', '$line-height', 'value', size]);`

File: src/kirby/directives/fit-header/fit-header.directive.ts [hunk 6]
```diff
@@ -0,0 +1,149 @@
+import { Directive, ElementRef, Input, OnDestroy, OnInit, Renderer2 } from '@angular/core';
+
+import { ResizeObserverService } from '@kirbydesign/designsystem/components/shared/resize-observer/resize-observer.service';
+import { ResizeObserverEntry } from '@kirbydesign/designsystem/components/shared/resize-observer/types/resize-observer-entry';
+
+declare var require: any;
+
+interface Size {
```
Comment (jakobe) [inline src/kirby/directives/fit-header/fit-header.directive.ts]: Consider naming:
```suggestion
interface HeadingSize {
```

File: src/kirby/directives/fit-header/fit-header.directive.ts [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+  name: string;
+  fontSize: string;
+  lineHeight: string;
+}
+
+export interface FitHeaderConfig {
```
Comment (jakobe) [inline src/kirby/directives/fit-header/fit-header.directive.ts]: Consider naming:
```suggestion
export interface FitHeadingConfig {
```

File: src/kirby/directives/fit-header/fit-header.directive.ts [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+  maxLines: number;
+}
+
+@Directive({
+  selector: `h1[kirbyFitHeader],h2[kirbyFitHeader],h3[kirbyFitHeader]`,
```
Comment (jakobe) [inline src/kirby/directives/fit-header/fit-header.directive.ts]: Consider naming:
```suggestion
  selector: `h1[kirbyFitHeading],h2[kirbyFitHeading],h3[kirbyFitHeading]`,
```

File: src/kirby/directives/fit-header/fit-header.directive.ts [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+})
+export class FitHeaderDirective implements OnInit, OnDestroy {
```
Comment (jakobe) [inline src/kirby/directives/fit-header/fit-header.directive.ts]: Consider naming:
```suggestion
export class FitHeadingDirective implements OnInit, OnDestroy {
```

File: src/kirby/directives/fit-header/fit-header.directive.ts [hunk 10]
```diff
@@ -0,0 +1,149 @@
...
+  // tslint:disable-next-line:no-input-rename
+  @Input('kirbyFitHeader') config?: FitHeaderConfig = {
```
Comment (jakobe) [inline src/kirby/directives/fit-header/fit-header.directive.ts]: Consider naming:
```suggestion
  @Input('kirbyFitHeading') config?: FitHeadingConfig = {
```

File: src/kirby/directives/fit-header/fit-header.directive.ts [hunk 11]
```diff
@@ -0,0 +1,149 @@
...
+    this.scalingHeader = true;
+
+    if (!this.clone) {
+      this.clone = this.generateClone();
+      this.renderer.appendChild(this.elementRef.nativeElement, this.clone);
+    }
+
+    this.renderer.setStyle(this.clone, 'width', `${this.elementRef.nativeElement.clientWidth}px`);
+
+    const fittedSize: Size =
```
Comment (jakobe) [inline src/kirby/directives/fit-header/fit-header.directive.ts]: Consider extracting to filter method to improve readability:
```
const fallbackSize = this.sizes[this.sizes.length - 1];
const fittedSize = this.sizes.find(this.canFitHeading.bind(this)) || fallbackSize;
...
}

private canFitHeading(size: Size) {
    this.setSize(this.clone, size);
    const lines = this.clone.clientHeight / parseInt(size.lineHeight);
    return lines <= this.config.maxLines;
  }
```

File: src/kirby/directives/fit-header/fit-header.module.ts [hunk 12]
```diff
@@ -0,0 +1,9 @@
+import { NgModule } from '@angular/core';
+
+import { FitHeaderDirective } from '@kirbydesign/designsystem/directives/fit-header/fit-header.directive';
+
+@NgModule({
+  declarations: [FitHeaderDirective],
+  exports: [FitHeaderDirective],
+})
+export class FitHeaderModule {}
```
Comment (jakobe) [inline src/kirby/directives/fit-header/fit-header.module.ts]: Consider moving to `PageModule` for now

File: src/kirby/components/page/page.component.ts [hunk 13]
```diff
@@ -92,12 +95,14 @@ export class PageActionsComponent {}
...
 })
-export class PageComponent implements OnInit, OnDestroy, AfterViewInit, AfterContentChecked {
+export class PageComponent
+  implements OnInit, OnDestroy, AfterViewInit, AfterContentChecked, OnChanges {
   @Input() title?: string;
   @Input() toolbarTitle?: string;
   @Input() titleAlignment?: 'left' | 'center' | 'right' = 'left';
   @Input() defaultBackHref?: string;
   @Input() hideBackButton?: boolean;
+  @Input() headerMaxLines?: number = 2;
```
[the remark this answers is not in the document]
Comment (bommariusser) [reply src/kirby/components/page/page.component.ts]: @jakobe @hafstad lad os sparke det i mål mandag. har svært at tage en beslutning før man har set det :) Men er sikker det er skide godt <3

File: src/kirby/directives/fit-heading/fit-heading.directive.ts [hunk 14]
```diff
@@ -0,0 +1,150 @@
...
+    this.isScalingHeader = false;
+  }
+
+  private canFitHeading(size: HeadingSize) {
+    this.setSize(this.clone, size);
+    const lines = this.clone.clientHeight / parseInt(size.lineHeight);
+    return lines <= this.config.maxLines;
+  }
+
+  private generateClone(): Element {
```
Comment (lydemann) [inline src/kirby/directives/fit-heading/fit-heading.directive.ts]: Maybe call this generateHostElementClone?

File: src/kirby/directives/fit-heading/fit-heading.directive.ts [hunk 15]
```diff
@@ -0,0 +1,150 @@
...
+
+@Directive({
+  selector: `h1[kirbyFitHeading],h2[kirbyFitHeading],h3[kirbyFitHeading]`,
+})
+export class FitHeadingDirective implements OnInit, OnDestroy {
+  // tslint:disable-next-line:no-input-rename
+  @Input('kirbyFitHeading') config?: FitHeadingConfig;
+
+  private previousWidth: number;
+  private clone: Element;
```
Comment (lydemann) [inline src/kirby/directives/fit-heading/fit-heading.directive.ts]: maybe call this hostElementClone

File: src/kirby/directives/fit-heading/fit-heading.directive.ts (hunk 14 above)
Comment (jakobe) [reply src/kirby/directives/fit-heading/fit-heading.directive.ts]: Good point, fixed 👍

File: src/kirby/directives/fit-heading/fit-heading.directive.ts (hunk 15 above)
Comment (jakobe) [reply src/kirby/directives/fit-heading/fit-heading.directive.ts]: Good point, fixed 👍

**Pull request merged**